[BUG] maybe_upcast_putmast also handle ndarray

makbigc opened the pull request:
-  closes #23823
-  2 tests added
-  passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
-  whatsnew entry

Follow h-vetinari's footstep. `maybe_upcast_putmask` was left untouched in #25425. Try to fix the bug so that `maybe_upcast_putmask` can also handle `ndarray` as well as `Series`.

gfyoung commented on pandas/tests/dtypes/cast/test_upcast.py:
```diff
@@ -0,0 +1,28 @@
+    expected = np.array([10, 61, 12])
+
+    result, _ = maybe_upcast_putmask(result, mask, other)
+    tm.assert_numpy_array_equal(result, expected)
```
I suspect these tests already have a home in another, existing test file.

gfyoung replied:
Also, reference the issue number as a comment under each test definition.

jreback commented on pandas/tests/dtypes/cast/test_upcast.py:
```diff
@@ -0,0 +1,28 @@
+from pandas.util import testing as tm
+
+
+def test_upcast_series():
```
can you parameterize these (both tests) as I suspect a myriad of cases (just a couple for now, e.g. int -> float you can start with)

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -209,7 +209,10 @@ def changeit():
-                    new_result = result.values.copy()
+                    if isinstance(result, np.ndarray):
+                        new_result = result.copy()
+                    else:
```
what other types are here? a Series?

this should be converted at a higher level (at the beginning of the function)

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -105,7 +105,7 @@ Performance Improvements
 ~~~~~~~~~
 
--
+- :func:`maybe_upcast_putmask` does not replace the element of the ``ndarray`` correctly (:issue:`23823`)
```
this doesn't warrant a whatsnew as its an internal impl detail

jreback replied:
can you remove

jreback reviewed (changes requested):
what problem is this solving?

makbigc commented:
I investigated further and found that I misunderstood the usage of `maybe_upcast_putmask`. 

The result array should be replaced with the element of other array starting from the beginning, not in the position of mask. This behavior is due to np.place

 `changeit` isn't called.
```
In [4]: arr1 = np.array([1,2,3])                                                

In [5]: other = np.array([61,62,63])                                            

In [6]: mask = np.array([False, True, False])                                   

In [7]: result, _ = maybe_upcast_putmask(arr1, mask, other)                     

In [8]: result                                                                  
Out[8]: array([ 1, 61,  3])
```

`changeit` is called.
```
In [24]: arr2 = np.array([1,2,3])                                                                                                           

In [25]: other2 = np.array([61.1, 62.2, 63.3])                                                                                              

In [26]: mask = np.array([False, True, False])                                                                                              

In [27]: result, _ = maybe_upcast_putmask(arr2, mask, other2)                                                                               

In [28]: result                                                                                                                             
Out[28]: array([ 1. , 61.1,  3. ])
```

`changeit` isn't called.
```
In [15]: arr2 = np.arange('2019-01-01', '2019-01-04', dtype='datetime64[D]')    

In [16]: other2 = np.arange('2018-01-01', '2018-01-04', dtype='datetime64[D]')  

In [17]: mask = np.array([False, True, False])                                  

In [18]: result, _ = maybe_upcast_putmask(arr2, mask, other2)                   

In [19]: result                                                                 
Out[19]: array(['2019-01-01', '2018-01-01', '2019-01-03'], dtype='datetime64[D]')
```
So the behavior for Series as an input is not correct. And `maybe_upcast_putmask` does not expect a series input.

`changeit` is called.
```
>>> s = pd.Series([10, 11, 12])
>>> t = pd.Series([np.nan, 61, np.nan])
>>> from pandas.core.dtypes.cast import maybe_upcast_putmask
>>> result, _ = maybe_upcast_putmask(s, np.array([False, True, False]), t)
>>> result  # correct
0    10
1    61
2    12
dtype: int64
```
Examples can be added into the function docstring to elucidate the usage. Should I do it in this PR?

jreback reviewed (changes requested):
can you update with your new tests. also you can do an assert on the input of the function itself (e.g. assert its an ndarray)

makbigc committed:
Add tests for maybe_upcast_putmask

makbigc committed:
Elaborate function usage, add examples and fix bug

makbigc committed:
Fix type mistake

makbigc commented:
When adding the tests, the following bug is found and fixed. 

```
In [4]: arr = np.arange(10, 15)                                                 

In [5]: other = np.array([61, 62])                                              

In [6]: mask = np.array([False, True, False, True, True])                       

In [7]: result, changed = maybe_upcast_putmask(arr, mask, other)                
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-7-3ec9bc405a70> in <module>
----> 1 result, changed = maybe_upcast_putmask(arr, mask, other)

~/Code/pandas/pandas/core/dtypes/cast.py in maybe_upcast_putmask(result, mask, other)
    242             else:
    243 
--> 244                 if isna(other[mask]).any():
    245                     return changeit()
    246 

IndexError: boolean index did not match indexed array along dimension 0; dimension is 2 but corresponding boolean dimension is 5
```

jreback commented on pandas/tests/dtypes/cast/test_upcast.py:
```diff
@@ -0,0 +1,62 @@
+from pandas.util import testing as tm
+
+
+def test_upcast_error():
```
are there more cases to consider? e.g. non-ndarray

jreback reviewed (changes requested):
minor comments, merge master, ping on green.

makbigc committed:
Add more cases into test_upcast_error

makbigc committed:
Merge master

jreback commented:
thanks @makbigc keep em coming!
